File: backend/src/vet/farmer-detail.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { In, Repository } from 'typeorm';
import { Vaccination, VaccinationStatus } from '../vaccinations/entities/vaccination.entity';
import { Flock } from '../flocks/entities/flock.entity';
import { User, UserRole } from '../users/entities/user.entity';
import { VetFarmerConnection, ConnectionStatus } from '../users/entities/vet-farmer-connection.entity';
import { SickReport, ReportStatus } from '../sick-reports/entities/sick-report.entity';

@Injectable()
export class FarmerDetailService {
  constructor(
    @InjectRepository(Vaccination)
    private vaccinationRepository: Repository<Vaccination>,

    @InjectRepository(Flock)
    private flockRepository: Repository<Flock>,

    @InjectRepository(User)
    private userRepository: Repository<User>,

    @InjectRepository(VetFarmerConnection)
    private connectionRepository: Repository<VetFarmerConnection>,

    @InjectRepository(SickReport)
    private sickReportRepository: Repository<SickReport>,
  ) {}

  /**
   * Get detailed information about a specific farmer/farm
   * @param vetPhone - The phone number of the logged-in vet
   * @param farmerId - The ID of the farmer to get details for
   */
  async getFarmerDetail(vetPhone: string, farmerId: number) {
    // Get vet user
    const vet = await this.userRepository.findOne({
      where: { phone: vetPhone, role: UserRole.VETERINARIAN },
    });

    if (!vet) {
      throw new Error('Veterinarian not found');
    }

    // Verify connection exists
    const connection = await this.connectionRepository.findOne({
      where: {
        vetId: vet.user_id,
        farmerId: farmerId,
        status: ConnectionStatus.ACCEPTED,
      },
    });

    if (!connection) {
      throw new Error('You are not connected to this farmer');
    }

    // Get farmer details
    const farmer = await this.userRepository.findOne({
      where: { user_id: farmerId },
    });

    if (!farmer) {
      throw new Error('Farmer not found');
    }

    // Get all flocks for this farmer
    const flocks = await this.flockRepository.find({
      where: { farmer: { user_id: farmer.user_id } },
      order: { created_at: 'DESC' },
    });

    // Get all vaccinations for these flocks administered by this vet
    const flockIds = flocks.map(f => f.flock_id);
    const vaccinations = flockIds.length > 0 ? await this.vaccinationRepository.find({
      where: {
        flock: In(flockIds),
        administered_by: {
          user_id: vet.user_id,
        },
      },
      relations: {
        flock: true,
        vaccine: true,
      },
      order: {
        date_given: 'DESC',
      },
    }) : [];

    // Get all sick reports for these flocks
    const sickReports = flockIds.length > 0 ? await this.sickReportRepository.find({
      where: {
        flockId: In(flockIds),
      },
      relations: {
        flock: true,
      },
      order: {
        created_at: 'DESC',
      },
    }) : [];

    // Calculate statistics
    const totalChickens = flocks.reduce((sum, flock) => sum + flock.bird_count, 0);
    const totalFlocks = flocks.length;
    const activeSickReports = sickReports.filter(sr => 
      sr.status !== ReportStatus.RESOLVED && sr.status !== ReportStatus.REVIEWED
    ).length;

    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const vaccinationsDue = vaccinations.filter(v => {
      if (v.status === VaccinationStatus.COMPLETED) return false;
      if (!v.next_due_date) return false;
      const nextDue = new Date(v.next_due_date);
      nextDue.setHours(0, 0, 0, 0);
      return nextDue >= today;
    }).length;

    // Build flock data
    const flockData = flocks.map(flock => ({
      flockId: flock.flock_id,
      batchName: flock.batch_name,
      birdCount: flock.bird_count,
      breed: flock.breed,
      createdAt: flock.created_at,
    }));

    // Build vaccination data
    const vaccinationData = vaccinations.map(v => ({
      vaccinationId: v.vaccination_id,
      flockId: v.flock.flock_id,
      flockName: v.flock.batch_name,
      vaccineName: v.vaccine.name_en,
      dateGiven: v.date_given,
      nextDueDate: v.next_due_date,
      status: v.status,
    }));

    // Build sick report data
    const sickReportData = sickReports.map(sr => ({
      reportId: sr.report_id,
      flockId: sr.flockId,
      flockName: sr.flock?.batch_name || 'Unknown',
      reportType: sr.reportType,
      affectedCount: sr.affectedCount,
      symptoms: sr.symptoms,
      reportDate: sr.reportDate,
      status: sr.status,
      createdAt: sr.created_at,
    }));

    return {
      farmer: {
        farmerId: farmer.user_id,
        name: farmer.name,
        phone: farmer.phone,
        village: farmer.village,
        province: farmer.province,
      },
      farmName: flocks.length > 0 ? flocks[0].batch_name : `${farmer.name}'s Farm`,
      summary: {
        totalChickens,
        totalFlocks,
        activeSickReports,
        vaccinationsDue,
      },
      flocks: flockData,
      vaccinations: vaccinationData,
      sickReports: sickReportData,
    };
  }
}

File: frontend/lib/services/farmer_detail_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';
import 'storage_service.dart';

class FarmerDetail {
  final FarmerInfo farmer;
  final String farmName;
  final FarmSummary summary;
  final List<FlockData> flocks;
  final List<VaccinationData> vaccinations;
  final List<SickReportData> sickReports;

  FarmerDetail({
    required this.farmer,
    required this.farmName,
    required this.summary,
    required this.flocks,
    required this.vaccinations,
    required this.sickReports,
  });

  factory FarmerDetail.fromJson(Map<String, dynamic> json) {
    return FarmerDetail(
      farmer: FarmerInfo.fromJson(json['farmer']),
      farmName: json['farmName'] ?? '',
      summary: FarmSummary.fromJson(json['summary']),
      flocks: (json['flocks'] as List)
          .map((flock) => FlockData.fromJson(flock))
          .toList(),
      vaccinations: (json['vaccinations'] as List)
          .map((vax) => VaccinationData.fromJson(vax))
          .toList(),
      sickReports: (json['sickReports'] as List)
          .map((report) => SickReportData.fromJson(report))
          .toList(),
    );
  }
}

class FarmerInfo {
  final int farmerId;
  final String name;
  final String phone;
  final String village;
  final String province;

  FarmerInfo({
    required this.farmerId,
    required this.name,
    required this.phone,
    required this.village,
    required this.province,
  });

  factory FarmerInfo.fromJson(Map<String, dynamic> json) {
    return FarmerInfo(
      farmerId: json['farmerId'] ?? 0,
      name: json['name'] ?? '',
      phone: json['phone'] ?? '',
      village: json['village'] ?? '',
      province: json['province'] ?? '',
    );
  }
}

class FarmSummary {
  final int totalChickens;
  final int totalFlocks;
  final int activeSickReports;
  final int vaccinationsDue;

  FarmSummary({
    required this.totalChickens,
    required this.totalFlocks,
    required this.activeSickReports,
    required this.vaccinationsDue,
  });

  factory FarmSummary.fromJson(Map<String, dynamic> json) {
    return FarmSummary(
      totalChickens: json['totalChickens'] ?? 0,
      totalFlocks: json['totalFlocks'] ?? 0,
      activeSickReports: json['activeSickReports'] ?? 0,
      vaccinationsDue: json['vaccinationsDue'] ?? 0,
    );
  }
}

class FlockData {
  final int flockId;
  final String batchName;
  final int birdCount;
  final String breed;
  final DateTime createdAt;

  FlockData({
    required this.flockId,
    required this.batchName,
    required this.birdCount,
    required this.breed,
    required this.createdAt,
  });

  factory FlockData.fromJson(Map<String, dynamic> json) {
    return FlockData(
      flockId: json['flockId'] ?? 0,
      batchName: json['batchName'] ?? '',
      birdCount: json['birdCount'] ?? 0,
      breed: json['breed'] ?? '',
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
    );
  }
}

class VaccinationData {
  final int vaccinationId;
  final int flockId;
  final String flockName;
  final String vaccineName;
  final DateTime dateGiven;
  final DateTime? nextDueDate;
  final String status;

  VaccinationData({
    required this.vaccinationId,
    required this.flockId,
    required this.flockName,
    required this.vaccineName,
    required this.dateGiven,
    this.nextDueDate,
    required this.status,
  });

  factory VaccinationData.fromJson(Map<String, dynamic> json) {
    return VaccinationData(
      vaccinationId: json['vaccinationId'] ?? 0,
      flockId: json['flockId'] ?? 0,
      flockName: json['flockName'] ?? '',
      vaccineName: json['vaccineName'] ?? '',
      dateGiven: DateTime.parse(json['dateGiven'] ?? DateTime.now().toIso8601String()),
      nextDueDate: json['nextDueDate'] != null ? DateTime.parse(json['nextDueDate']) : null,
      status: json['status'] ?? '',
    );
  }
}

class SickReportData {
  final int reportId;
  final int flockId;
  final String flockName;
  final String reportType;
  final int affectedCount;
  final String symptoms;
  final String reportDate;
  final String status;
  final DateTime createdAt;

  SickReportData({
    required this.reportId,
    required this.flockId,
    required this.flockName,
    required this.reportType,
    required this.affectedCount,
    required this.symptoms,
    required this.reportDate,
    required this.status,
    required this.createdAt,
  });

  factory SickReportData.fromJson(Map<String, dynamic> json) {
    return SickReportData(
      reportId: json['report_id'] ?? json['reportId'] ?? 0,
      flockId: json['flock_id'] ?? json['flockId'] ?? 0,
      flockName: json['flockName'] ?? json['batch_name'] ?? 'Unknown',
      reportType: json['reportType'] ?? '',
      affectedCount: json['affectedCount'] ?? 0,
      symptoms: json['symptoms'] ?? '',
      reportDate: json['reportDate'] ?? json['report_date'] ?? '',
      status: json['status'] ?? '',
      createdAt: DateTime.parse(json['createdAt'] ?? json['created_at'] ?? DateTime.now().toIso8601String()),
    );
  }
}

class FarmerDetailService {
  Future<FarmerDetail> getFarmerDetail(int farmerId) async {
    final token = await StorageService.getToken();

    if (token == null || token.isEmpty) {
      throw Exception('Authentication token is missing. Please log in again.');
    }

    final response = await http.get(
      Uri.parse('${ApiConfig.baseUrl}/vet/farmer/$farmerId'),
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'Authorization': 'Bearer $token',
      },
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return FarmerDetail.fromJson(data);
    } else {
      final message = response.body.isNotEmpty
          ? response.body
          : 'Failed to load farmer details';
      throw Exception('Error ${response.statusCode}: $message');
    }
  }
}